OrbitVulkanLayer/SubmissionTrackerTest.cpp

```diff
@@ -0,0 +1,1446 @@
+  tracker_.MarkDebugMarkerBegin(command_buffer_, "Marker", {});
+  tracker_.MarkDebugMarkerEnd(command_buffer_);
+}
+
+TEST_F(SubmissionTrackerTest, CanRetrieveDebugMarkerTimestampsForACompleteSubmission) {
+  ExpectFourNextReadyQuerySlotCalls();
+  EXPECT_CALL(dispatch_table_, GetQueryPoolResults)
+      .WillRepeatedly(Return(mock_get_query_pool_results_function_all_ready_));
+  std::vector<uint32_t> actual_reset_slots;
+  EXPECT_CALL(timer_query_pool_, ResetQuerySlots)
+      .Times(1)
+      .WillOnce(SaveArg<1>(&actual_reset_slots));
+  orbit_grpc_protos::CaptureEvent actual_capture_event;
+  auto mock_enqueue_capture_event =
+      [&actual_capture_event](orbit_grpc_protos::CaptureEvent&& capture_event) {
+        actual_capture_event = std::move(capture_event);
+        return true;
+      };
+
+  const char* text = "Text";
+  constexpr uint64_t expected_text_key = 111;
+  auto mock_intern_string_if_necessary_and_get_key = [&text](std::string str) {
+    EXPECT_STREQ(text, str.c_str());
+    return expected_text_key;
+  };
+  EXPECT_CALL(*producer_, InternStringIfNecessaryAndGetKey)
+      .Times(1)
+      .WillOnce(Invoke(mock_intern_string_if_necessary_and_get_key));
+  EXPECT_CALL(*producer_, EnqueueCaptureEvent)
+      .Times(1)
+      .WillOnce(Invoke(mock_enqueue_capture_event));
+
+  Color expected_color{1.f, 0.8f, 0.6f, 0.4f};
+
+  producer_->StartCapture();
+  tracker_.TrackCommandBuffers(device_, command_pool_, &command_buffer_, 1);
+  tracker_.MarkCommandBufferBegin(command_buffer_);
+  tracker_.MarkDebugMarkerBegin(command_buffer_, text, expected_color);
+  tracker_.MarkDebugMarkerEnd(command_buffer_);
+  tracker_.MarkCommandBufferEnd(command_buffer_);
+  pid_t tid = GetCurrentThreadId();
+  uint64_t pre_submit_time = MonotonicTimestampNs();
+  std::optional<QueueSubmission> queue_submission_optional =
+      tracker_.PersistCommandBuffersOnSubmit(1, &submit_info_);
+  tracker_.PersistDebugMarkersOnSubmit(queue_, 1, &submit_info_, queue_submission_optional);
+  uint64_t post_submit_time = MonotonicTimestampNs();
+  tracker_.CompleteSubmits(device_);
+
+  EXPECT_THAT(actual_reset_slots,
+              UnorderedElementsAre(kSlotIndex1, kSlotIndex2, kSlotIndex3, kSlotIndex4));
+  EXPECT_TRUE(actual_capture_event.has_gpu_queue_submission());
+  const orbit_grpc_protos::GpuQueueSubmission& actual_queue_submission =
+      actual_capture_event.gpu_queue_submission();
+  EXPECT_EQ(actual_queue_submission.num_begin_markers(), 1);
+  EXPECT_EQ(actual_queue_submission.completed_markers_size(), 1);
+  const orbit_grpc_protos::GpuDebugMarker& actual_debug_marker =
+      actual_queue_submission.completed_markers(0);
+
+  ExpectDebugMarkerEndEq(actual_debug_marker, kTimestamp3, expected_text_key, expected_color, 0);
+  ExpectDebugMarkerBeginEq(actual_debug_marker, kTimestamp2, pre_submit_time, post_submit_time,
+                           tid);
+}
+
+TEST_F(SubmissionTrackerTest, CanRetrieveDebugMarkerEndEvenWhenNotCapturedBegin) {
```

Comment:
As above, I guess "when begin not captured" is better.

Reply:
Done.